### PR TITLE
Remove gui globals

### PR DESCRIPTION
**Pull request by tobiasdiez**
Removes the class GUIGlobals, whose members where not used any more. Also cleaned up the preferences and deleted some unused variables. Finally, marked Globals as deprecated. 

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not: Issue created at <https://github.com/JabRef/user-documentation/issues>.

### CODE REVIEW

File: src/main/java/org/jabref/gui/icon/IconTheme.java [hunk 1]
```diff
@@ -110,7 +108,7 @@ public static URL getIconUrl(String name) {
         Map<String, String> result = new HashMap<>();
 
         try (BufferedReader in = new BufferedReader(
-                                                    new InputStreamReader(url.openStream(), StandardCharsets.ISO_8859_1))) {
+                new InputStreamReader(url.openStream(), StandardCharsets.ISO_8859_1))) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/icon/IconTheme.java]: Any reason this is not utf 8?

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/icon/IconTheme.java]: No idea...

**Comment by credmond** [comment]
@tobiasdiez: What's the ultimate plan or decision, for Globals for things that do not belong in StateManager} or  PreferencesService...? Also how does this to in to DefaultInjector the relies heavily on it?

**Comment by tobiasdiez** [comment]
The long-term goal is to completely remove the `Globals` class. Instead, the usage of it is slowly replaced by dependency injection (with the DefaultInjector being the link between the DI and `Globals` by delegating to the later). Once a property is no longer accessed via `Globals`, it can be removed there and is instead created in the  `DefaultInjector`.

**Comment by credmond** [comment]
So if a class -- any class anywhere -- needed something, it'd use Injector.instantiateModelOrService(Whatever.class);... that's basically the replacement pattern you're referring to...correct?

Regarding the responsibility for "what" builds an object initially, I am curious.

Pick JournalAbbreviationRepository for example.

Today, Launcher asks JournalAbbreviationLoader to build that, and then sticks the results into Globals.journalAbbreviationRepository and DependencyInjector refers to Globals.journalAbbreviationRepository when it's requested via DI.

Could you give a brief explanation about how you would re-factor that, as an example (I'm just curious). I'm curious where the responsibility for actually instantiating and populating a class like JournalAbbreviationRepository is, and at what point during startup (or first access) it occurs.

Today DefaultInjector is quite dumb -- it doesn't know how to build anything. But would it ultimately be responsible? Or would another class be responsible for building it and then telling DefaultInjector (which is basically what it is today, albeit a bit messy).

**Comment by credmond** [comment]
Two more questions; I see a good few changes recently where people are handing around, for example, PreferencesService, in a new constructor parameter -- the commit messages are talking about the change being related the reliance on Globals. But they're using constructor args, instead of DefaultInjector. What's the pattern on when to use what approach? Obviously both do the same job, but... what's the decision making process there and why?

This leads me to my second question, regarding testing and mocking. When writing a test, it's much easier to mock an object and use that in a class's constructor when constructing the class you want to test. With this DI framework, I'm not sure how mocking would work, if at all -- e.g., if the class you're testing uses Injector.instantiateModelOrService to get dependencies, it would always get a concrete dependency and therefore it's a) not really a "unit" test, as there's at least 2 or more classes involved, and b) harder to test without the benefits of mocking.

Any thoughts?

**Comment by tobiasdiez** [comment]
> Pick JournalAbbreviationRepository for example.
> Could you give a brief explanation about how you would re-factor that, as an example (I'm just curious). I'm curious where the responsibility for actually instantiating and populating a class like JournalAbbreviationRepository is, and at what point during startup (or first access) it occurs.

Step 1: Replace all references of `Globals.journalAbbreviationRepository` by passing it down via constructors (e.g. https://github.com/JabRef/jabref/blob/2a48a8c9a77cdc25119ffb5b71fc439a798bcc1a/src/main/java/org/jabref/gui/journals/AbbreviateAction.java#L132 should use https://github.com/JabRef/jabref/blob/2a48a8c9a77cdc25119ffb5b71fc439a798bcc1a/src/main/java/org/jabref/gui/journals/AbbreviateAction.java#L53 - this one is easy, most often you have to introduce a whole chain of such constructor args)
Step 2: Remove `Globals.journalAbbreviationRepository`
Step 3: Replace https://github.com/JabRef/jabref/blob/2a48a8c9a77cdc25119ffb5b71fc439a798bcc1a/src/main/java/org/jabref/gui/DefaultInjector.java#L39-L40 by what is currently in https://github.com/JabRef/jabref/blob/2a48a8c9a77cdc25119ffb5b71fc439a798bcc1a/src/main/java/org/jabref/cli/Launcher.java#L170-L171


> Today DefaultInjector is quite dumb -- it doesn't know how to build anything. But would it ultimately be responsible? Or would another class be responsible for building it and then telling DefaultInjector (which is basically what it is today, albeit a bit messy).

`DefaultInjector` would be responsible for coordinating all the constructions, either by directly calling the constructor or outsourcing it to a loader class if its more complicated.

> e.g., if the class you're testing uses Injector.instantiateModelOrService to get dependencies, it would always get a concrete dependency

You can replace the injector, but yeah in general it is more complicated to mock dependencies that way. That's also why we almost always use the constructor injection method.

**Comment by calixtus** [comment]
Hi, we are currently discussing different DI approaches and frameworks. Our long term goal is to completly remove globals and also DefaultInjector and replace it with a more widely supported and easier to use DI framework. But we are still evaluating how this affects the currently used afterburnerfx package we are using for the mvvm pattern. Currently Globals also works as some kind of cache, since (in the example above) calling loadRepository is quite expensive. So this is something we have to keep in mind too.

From time to time we create some prs replacing calls to Globals with constructor injection, but as you already mentioned, this is a bit frustrating, since the calls are so very deeply buried.

Currently this is one of multiple long term projects in the background to modernize the code. Others would be to use streams instead of the obsolete EventBus and using the command pattern to finally fix the broken UndoRedo, modularize the logic and the gui of jabref to provide a separate cli only thin package of jabref.

**Comment by calixtus** [comment]
About testing statics like Injector: See BackupFileUtilTest for an example. Could be customized for Injector.
Also about DefaultInjector: I wonder if we can't replace the DefaultInjector with calls to Injector.setModelOrService() ...

**Comment by tobiasdiez** [comment]
> Our long term goal is to completly remove globals and also DefaultInjector and replace it with a more widely supported and easier to use DI framework. But we are still evaluating how this affects the currently used afterburnerfx package we are using for the mvvm pattern. 

You can just replace the calls to `Injector` in `DefaultInjector` with whatever method your favorite DI is providing to init a new class. Then you are only using afterburner to internally wire this `PresenterFactory` to the `FXMLLoader` that is used to create the view from the fxml file. Should be fairly straightforward.

**Comment by credmond** [comment]
> About testing statics like Injector: See BackupFileUtilTest for an example. Could be customized for Injector. Also about DefaultInjector: I wonder if we can't replace the DefaultInjector with calls to Injector.setModelOrService() ...

Injector.setModelOrService(); does work! Basically, setting this maintains different map that afterburner will consult by default (0 size by default), anytime a lookup is formed -- and only if it's not there, will it consult the supplier (which isDefaultInjector.createDependency()). So that's good news in the short-term, for testing purposes anyway.

**Pull request merged**